src/EditorFeatures/CSharpTest/UseRecursivePatterns/UseRecursivePatternsTests.cs

```diff
@@ -0,0 +1,239 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.CSharp.UseRecursivePatterns;
+using Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.CodeRefactorings;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.UseRecursivePatterns
+{
+    public partial class UseRecursivePatternsTests : AbstractCSharpCodeActionTest
+    {
+        protected override CodeRefactoringProvider CreateCodeRefactoringProvider(Workspace workspace, TestParameters parameters)
+           => new CSharpUseRecursivePatternsCodeRefactoringProvider();
+
+        [Fact]
+        public async Task Test1()
+        {
+            await TestInRegularAndScriptAsync(
+@"class C
+{
+    bool M()
+    {
+        return b.P1 == 1 [||]&& b.P2 == 2;
+    }
+}",
+@"class C
+{
+    bool M()
+    {
+        return b is
+        {
+            P1: 1, P2: 2
+        };
+    }
+}");
+        }
+
+        [Fact]
+        public async Task Test2()
+        {
+            await TestInRegularAndScriptAsync(
+@"class C
+{
+    bool M()
+    {
+        return b.MetadataName == string.Empty &&
+            b.ContainingType == null &&
+            b.ContainingNamespace != null &&
+            b.ContainingNamespace.Name == string.Empty [||]&&
+            b.ContainingNamespace.ContainingNamespace == null;
+    }
+}",
+@"class C
+{
+    bool M()
+    {
+        return b is
+        {
+            MetadataName: string.Empty, ContainingType: null, ContainingNamespace:
+            {
+                Name: string.Empty, ContainingNamespace: null
+            }
+        };
+    }
+}");
+        }
+
+        [Fact]
+        public async Task Test3()
+        {
+            await TestInRegularAndScriptAsync(
+@"class C
+{
+    bool M()
+    {
+        return type.Name == ""ValueTuple"" &&
+            type.ContainingSymbol is var declContainer &&
+            declContainer.Kind == SymbolKind.Namespace [||]&&
+            declContainer.Name == ""System"";
+    }
+}",
+@"class C
+{
+    bool M()
+    {
+        return type is
+        {
+            Name: ""ValueTuple"", ContainingSymbol:
+            {
+                Name: ""System"", Kind: SymbolKind.Namespace
+            }
+
+            declContainer
```

Comment:
what is this random name?

Reply:
this is the `var declContainer` pattern in "before" which has become the pattern designation.

Reply:
this reads super weirdly.  especially with the multiple newlines, and no clear indication about which part of hte pattern this corresponds to.

Reply:
how would you expect this to be formatted (a multi-line property pattern + designation)?

Reply:
another possibility: type + multi-line property pattern + designation

Reply:
what does 'declContainer' associate with here?  is it the designator for `ContainingSymbol: { ... } declContainer`?

Reply:
yeah having both a newline and a blank line here is bad.  it looks likke declContainer is completed separated from teh pattern.  We really shoudl not do this.  It should at least be:

```c#
return type is
        {
            Name: ""ValueTuple"",
            ContainingSymbol: { Name: ""System"", Kind: SymbolKind.Namespace } declContainer
        }
```

However, since declContainre isn't used, we should remove it as well.

Reply:
Could we modify formatting engine to NOT insert new lines in certain places like, e.g

- an empty property pattern clause `{}` does not need to break in multiple lines
- we should not insert a new line before a variable designation ever

Or more rules? lookin at tests, only these two would result in a way better source text.

Reply:
Also, I'm not sure if we should deal with indentation when breaking lists into multiple lines.

I think unless we want to explicitly perverse trivia, we don't need to do anything on formatting in this feature. These all should go to formatting rules, right?

Reply:
1. if you are not preserving trivia and you are using ElasticTrivia, then yes, you just need to update formatting rules to do something sensible there.
2. if you are preserving trivia, then it's likely you'll need to write some custom logic so that hte preserved trivia still looks good.

I'm fine with any approach you take that ends up with a good final result for the user :)  '1' would be a good place to start from to see how far that gets us.